### PR TITLE
Add random superusers

### LINKED ISSUE #8980: Tests that involve FxA are failing on circleci integration-test job.
Some tests using fxa are failing due to improper teardown within a plugin we use. We have found a fix and filled an issue with the plugin [here](https://github.com/mozilla/pytest-fxa/issues/15).

For now I will add a ```wget``` command that will delete all accounts matching ```uitest``` within the restmail DB.

### PR DESCRIPTION
**Pull request by jrbenny35**
This should fix #8980.

* Changed fxa account generation to allow assigning an fxa ID to a created user.
* Moved pytest execution to makefile to access variables.
* Removed second login tests as the devhub tests also log in.

### CODE REVIEW

File: requirements/uitests.txt [hunk 1]
```diff
@@ -71,8 +71,9 @@ zope.interface==4.5.0 \
     --hash=sha256:e6b77ae84f2b8502d99a7855fa33334a1eb6159de45626905cb3e454c023f339 \
     --hash=sha256:a1daf9c5120f3cc6f2b5fef8e1d2a3fb7bbbb20ed4bfdc25bc8364bc62dcf54b \
     --hash=sha256:57c38470d9f57e37afb460c399eb254e7193ac7fb8042bd09bdc001981a9c74c
-pluggy==0.6.0 \
-    --hash=sha256:7f8ae7f5bdf75671a718d2daf0a64b7885f74510bcd98b1a0bb420eb9a9d0cff
+    pluggy==0.7.1 \
+    --hash=sha256:6e3836e39f4d36ae72840833db137f7b7d35105079aee6ec4a62d9f80d594dd1 \
+    --hash=sha256:95eb8364a4708392bae89035f45341871286a333f749c3141c20573d2b3876e1
```
Comment (EnTeQuAk) [inline requirements/uitests.txt]: indentation :scream:

Comment (EnTeQuAk) [reply requirements/uitests.txt]: Let's back out this change maybe with a rebase to master?

Comment (jrbenny35) [reply requirements/uitests.txt]: Master doesn't have this updated. I'll get a mismatch error.

File: Makefile-docker [hunk 2]
```diff
@@ -1,3 +1,5 @@
+export FXA_EMAIL=uitest-$(shell uuid)@restmail.net
+export FXA_PASSWORD=uitester
```
Comment (EnTeQuAk) [inline Makefile-docker]: Let's add a comment what this is used for. Maybe even try to put this directly into the `ui-test` action but I'm not sure if make allows this.

File: Makefile-docker [hunk 3]
```diff
@@ -156,13 +158,13 @@ setup-ui-tests:
...
 	python manage.py waffle_switch activate-autograph-signing on --create
-	python manage.py createsuperuser --email=uitest@restmail.net --username=uitest --noinput --add-to-supercreate-group --hostname=olympia-frontend.test
 
 ui-tests: setup-ui-tests
 	# Generate test add-ons and force a reindex to make sure things are updated
+	pip install --progress-bar=off --no-deps -r requirements/uitests.txt
 	python manage.py generate_ui_test_addons
 	python manage.py reindex --force --noinput --wipe
-	pip install --progress-bar=off --no-deps -r requirements/uitests.txt
+	pytest --driver Firefox -n 2 tests/ui/
```
Comment (EnTeQuAk) [inline Makefile-docker]: Parallel tests! Did you test if we get an actual performance benefit here?

File: Makefile-docker [hunk 4]
```diff
@@ -156,13 +158,13 @@ setup-ui-tests:
 	# Create a proper superuser that can be used to access the API
 	python manage.py waffle_switch super-create-accounts on --create
 	python manage.py waffle_switch activate-autograph-signing on --create
-	python manage.py createsuperuser --email=uitest@restmail.net --username=uitest --noinput --add-to-supercreate-group --hostname=olympia-frontend.test
 
 ui-tests: setup-ui-tests
```
Comment (EnTeQuAk) [inline Makefile-docker]: As shown in my recent change in your branch, let's name this `run-ui-tests` to make this action more explicit.

File: docker-compose.yml [hunk 5]
```diff
@@ -94,12 +94,13 @@ services:
   addons-frontend:
     <<: *env
     environment:
+      - HOSTNAME=uitests
```
Comment (EnTeQuAk) [inline docker-compose.yml]: Why is this needed?

File: src/olympia/landfill/serializers.py [hunk 6]
```diff
@@ -28,10 +35,53 @@
...
+        session = fxa_client.create_account(fxa_account.email,
+                                            fxa_account.password)
+        account.fetch()
+        message = account.wait_for_email(
+            lambda m: 'x-verify-code' in m['headers'] and
+            session.uid == m['headers']['x-uid']
+        )
+        session.verify_email_code(message['headers']['x-verify-code'])
+        self.fxa_id = session.uid
+        log.debug("fxa account created: {}".format(fxa_account))
```
Comment (EnTeQuAk) [inline src/olympia/landfill/serializers.py]: I really hate side-effects in classes like that (we have them everywhere, and they're always a pain to debug).

Let's pass `fxa_password` and `fxa_email` as arguments to `_create_fxa_user` and `_create_addon_user` as well as pass the `fxa_id` to `_create_addon_user` by returning it from `_create_fxa_user`.

```python
def __init__(self):
    fxa_id = _create_fxa_user(username, password)
    self.user = _create_addon_user(fxa_id)
```

We still have that side-effect of setting `self.user` but I'd say that's okay and is how this class generally works.

File: src/olympia/users/models.py [hunk 7]
```diff
@@ -87,25 +87,30 @@ def widget_attrs(self, widget):
...
         # an anonymous username.
         now = timezone.now()
         user = self.model(
-            username=username, email=email, fxa_id=fxa_id,
-            last_login=now)
+            username=username,
+            email=email,
+            fxa_id=fxa_id,
+            last_login=now,
+            **kwargs
```
Comment (EnTeQuAk) [inline src/olympia/users/models.py]: Did you add this to support setting `display_name`?

File: src/olympia/users/models.py [hunk 8]
```diff
@@ -87,25 +87,30 @@ def widget_attrs(self, widget):
...
         if username is None:
             user.anonymize_username()
         log.debug('Creating user with email {} and username {}'.format(
             email, username))
         user.save(using=self._db)
         return user
 
-    def create_superuser(self, username, email, fxa_id=None):
+    def create_superuser(
+            self, username, email, fxa_id=None):
```
Comment (EnTeQuAk) [inline src/olympia/users/models.py]: I guess line break not needed anymore?

File: src/olympia/users/models.py [hunk 9]
```diff
@@ -87,25 +87,30 @@ def widget_attrs(self, widget):
...
         """
         Creates and saves a superuser.
         """
-        user = self.create_user(username, email, fxa_id)
+        user = self.create_user(username, email, fxa_id=fxa_id)
```
Comment (EnTeQuAk) [inline src/olympia/users/models.py]: Then start being explicit and make this `user = self.create_user(username=username, email=email, fxa_id=fxa_id)` which usually helps readability anyway.

File: tests/ui/conftest.py [hunk 10]
```diff
@@ -77,10 +69,17 @@ def selenium(selenium, request):
 
 
 @pytest.fixture
-def devhub_login(selenium, fxa_account):
+def fxa_account():
+    setattr(fxa_account, 'email', os.environ['FXA_EMAIL'])
+    setattr(fxa_account, 'password', os.environ['FXA_PASSWORD'])
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: What's wrong with doing

```python
fxa_account.email = os.environ['FXA_EMAIL']
fxa_account.password = os.environ['FXA_PASSWORD']
```

Also, we should fail in a reasonable way in case `FXA_EMAIL` and `FXA_PASSWORD` are not set in the environment.

File: tests/ui/conftest.py [hunk 11]
```diff
@@ -104,8 +103,8 @@ def pytest_configure(config):
     prefix_indexes(config)
 
 
-@pytest.fixture(scope='session')
-def es_test(pytestconfig):
+@pytest.fixture
+def es_test(pytestconfig, db):
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: Why did this change? We shouldn't enable database access to all elasticsearch tests by default.

File: Makefile-docker (hunk 3 above)
Comment (jrbenny35) [reply Makefile-docker]: Yeah its a good bit faster. Didn't get any hard numbers though.

File: docker-compose.yml (hunk 5 above)
Comment (jrbenny35) [reply docker-compose.yml]: For the node_config package that loads node configs on addons-frontend.

File: src/olympia/users/models.py (hunk 7 above)
Comment (jrbenny35) [reply src/olympia/users/models.py]: Yes

File: src/olympia/users/models.py (hunk 8 above)
Comment (jrbenny35) [reply src/olympia/users/models.py]: No, I don't know why there was a break.

File: tests/ui/conftest.py (hunk 10 above)
Comment (EnTeQuAk) [reply tests/ui/conftest.py]: `setattr` doesn't do what i think you think it does. The value you're passing in isn't a default value, it's the value this object gets set to.

```
Signature: setattr(obj, name, value)
Docstring:
Sets the named attribute on the given object to the specified value.

setattr(x, 'y', v) is equivalent to ``x.y = v''
```

It's literally the same.

Comment (jrbenny35) [reply tests/ui/conftest.py]: Oh whoops, I may have been thinking of something else.

File: src/olympia/landfill/serializers.py [hunk 12]
```diff
@@ -64,23 +62,23 @@ def _create_fxa_user(self):
...
                 display_name='uitest'
             )
         except Exception as e:
-            log.debug("There was a problem creating the user: {}."
-                " Returning user from database".format(e)
-            )
-            self.user = UserProfile.objects.get(username='uitest')
+            log.debug('There was a problem creating the user: {}.'
+                      ' Returning user from database'.format(e)
+                      )
```
Comment (EnTeQuAk) [inline src/olympia/landfill/serializers.py]: The parenthesis is wrong here, should either be at the beginning of the line or ideally behind the `.format(e)`

**Comment by jrbenny35** [comment]
Using ```-n 2``` sometimes results in a ```Webdriver: Connection Refused``` error. Is parallel worth it? We may get failures that are just errors due to geckodriver race conditions.

**Comment by EnTeQuAk** [comment]
> Using -n 2 sometimes results in a Webdriver: Connection Refused error. Is parallel worth it? We may get failures that are just errors due to geckodriver race conditions.

I don't think it's worth it. Tests that run predictably are worth the time imho, we can optimize that later but I'd start with something that reliably works.

File: src/olympia/landfill/serializers.py (hunk 12 above)
Comment (jrbenny35) [reply src/olympia/landfill/serializers.py]: Flake8 doesn't like it at the beginning.

File: src/olympia/users/models.py (hunk 8 above)
Comment (EnTeQuAk) [reply src/olympia/users/models.py]: Can we get rid of this line-break before merging please?

File: tests/ui/conftest.py [hunk 13]
```diff
@@ -69,18 +61,34 @@ def selenium(selenium, request):
...
+
+    Returns the email and password of the fxa account set in Makefile-docker.
+
+    """
+    try:
+        fxa_account.email = os.environ['FXA_EMAIL']
+        fxa_account.password = os.environ['FXA_PASSWORD']
+    except KeyError:
+        if request.node.get_marker('fxa_login'):
+            pytest.skip('Skipping test because no fxa account was found.')
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: It may make sense to add something like "Are FXA_EMAIL and FXA_PASSWORD environment variables set?" to the message since in our case it is most likely an error and unexpected.
